Ensure get_default_reader_themes returns an array and sets the array as a class attribute, and handle failed themes_api in UI

johnwatkins0 opened the pull request:
## Summary

As the title describes, this makes a couple of small changes to the `ReaderThemes::get_default_reader_themes` method to ensure it (a) always returns an array regardless of the result of `themes_api` and (b) always sets that array as the `$default_reader_themes` class property. This latter point ensures `get_default_reader_themes` will never fully execute twice for a given instance.

This should solve https://wordpress.org/support/topic/issue-during-activating-the-updated-plugins/#post-13383737 

<!-- Please reference the issue this PR addresses. -->
Fixes #5367

## New addition to PR

I've also handled a failed_api response in the UI. Instead of letting `get_default_reader_themes` silently fail when `themes_api`'s response is invalid, it now returns an error. 

Where `get_default_reader_themes` is called, if the result is an error, that error is added to the `ReaderThemes` object state. When retrieving reader themes, the REST controller then checks for that error, and if there is an error, it sets a header on the response. In the JS app, we check for that header. If the header exists, we display a notice indicating that there is a problem.

### After
Settings screen: 
<img width="1065" alt="Screen Shot 2020-09-16 at 8 29 27 PM" src="https://user-images.githubusercontent.com/9020968/93408983-795be780-f85b-11ea-8028-7a2b29fe6db9.png">

Wizard:
<img width="932" alt="Screen Shot 2020-09-16 at 8 29 44 PM" src="https://user-images.githubusercontent.com/9020968/93408994-7e209b80-f85b-11ea-8bcc-32fdf8f6e3e4.png">


## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Ensure get_default_reader_themes returns an array and sets the array as a class attribute

pierlon commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +209,21 @@ public function get_default_reader_themes() {
+			 *
+			 * @see https://wordpress.org/support/topic/issue-during-activating-the-updated-plugins/#post-13383737
+			 */
+			if ( is_wp_error( $response ) || ! is_object( $response ) || ! is_array( $response->themes ) ) {
```
You would need to check if the `themes` property exists as well before accessing it, otherwise a `Undefined property` notice would occur.

```suggestion
			if (
				is_wp_error( $response )
				|| ! is_object( $response )
				|| ! property_exists( $response, 'themes' )
				|| ! is_array( $response->themes )
			) {
```

johnwatkins0 committed:
Check that property exists before accessing it

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter commented:
I'm trying to reproduce the issue by adding a filter that does:

```php
add_filter( 'themes_api_result', '__return_null' );
```

When accessing the settings screen:

> ![image](https://user-images.githubusercontent.com/134745/92984757-7a5ad680-f461-11ea-8ae3-77f7d65c09f0.png)

This are the errors:

```
Notice: Trying to get property 'themes' of non-object in /app/public/content/plugins/amp/src/Admin/ReaderThemes.php on line 233
Warning: array_filter() expects parameter 1 to be array, null given in /app/public/content/plugins/amp/src/Admin/ReaderThemes.php on line 236
Warning: array_map(): Expected parameter 2 to be an array, null given in /app/public/content/plugins/amp/src/Admin/ReaderThemes.php on line 240
Warning: Invalid argument supplied for foreach() in /app/public/core-dev/src/wp-includes/class-wp-list-util.php on line 157
```

This aligns with the what is reported on the support topic:

```
PHP message: PHP Warning: array_filter() expects parameter 1 to be array, null given in /home/zodev1zo/webapps/dev1/wp-content/plugins/amp/src/Admin/ReaderThemes.php on line 236
PHP message: PHP Warning: array_map(): Expected parameter 2 to be an array, null given in /home/zodev1zo/webapps/dev1/wp-content/plugins/amp/src/Admin/ReaderThemes.php on line 240
PHP message: PHP Warning: Invalid argument supplied for foreach() in /home/zodev1zo/webapps/dev1/wp-includes/class-wp-list-util.php on line 151
PHP message: PHP Fatal error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 1511424 bytes) in /home/zodev1zo/webapps/dev1/wp-includes/wp-db.php on line 1989
```

When I check out this branch then I don't get those notices or warnings:

> ![image](https://user-images.githubusercontent.com/134745/92984821-1389ed00-f462-11ea-8496-5f6e4d20067f.png)

Nevertheless, there are no themes presented. I suppose that is expected, although the user won't know why they don't have any themes presented to them.

Should there be some notice that tells the user that there appears to be a problem connecting to WordPress.org? Should the REST API return an error state in this case that is shown in an error notice?

westonruter commented:
> This aligns with the what is reported on the support topic:

Er, except there's a fatal error being reported regarding memory usage on the support topic. Not sure why.

johnwatkins0 commented:
> Should there be some notice that tells the user that there appears to be a problem connecting to WordPress.org? Should the REST API return an error state in this case that is shown in an error notice?

I was thinking the same but thought we might want follow up in a separate issue. It shouldn't be too involved, though, so I'll add it to this branch

johnwatkins0 committed:
Reader themes: handle error fromr wp.org themes_api failure

johnwatkins0 committed:
Merge branch 'fix/5367-themes-api-result-sanitization' of https://github.com/ampproject/amp-wp into fix/5367-themes-api-result-sanitization

johnwatkins0 committed:
Remove debug code

johnwatkins0 committed:
Reader themes: restore returned WP_Error after merging from remote

johnwatkins0 committed:
In PHPUnit test, delete related transient when test instance is re-created on the fly

johnwatkins0 committed:
Settings/onboarding: break themes API error notice into separate reusable component

johnwatkins0 committed:
Remove unused import

johnwatkins0 commented:
@pierlon @westonruter I have pushed updates to handle a failed `themes_api` in the settings and wizard UI. I've also updated the PR description, and added QA steps to the issue. Re-requesting review.

johnwatkins0 committed:
Fix case of error code

johnwatkins0 committed:
Add test for themes_api success

johnwatkins0 committed:
Remove extra slashes

johnwatkins0 commented:
As discussed on a call today, I'm going to update to handle the case where the themes array exists in the `themes_api` response but is empty. In this case, we'll display a different message indicating that a plugin is probably causing this. I'll also look for a noncomplicated way to cache the the empty-array result in a manner allowing it to be invalidated when there is a change to the active plugins.

westonruter commented:
> I'll also look for a noncomplicated way to cache the the empty-array result in a manner allowing it to be invalidated when there is a change to the active plugins.

I just noticed that every time you access a tab on `/wp-admin/theme-install.php` that a `themes_api()` call is made. So I don't think we need to worry about caching the empty result.

johnwatkins0 committed:
Update handling of errors from themes_api: handle empty array

johnwatkins0 committed:
Update unit tests for themes_api failures

johnwatkins0 committed:
Add missing docblock comment

johnwatkins0 committed:
Remove unused import

schlessera commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,38 @@ public function get_default_reader_themes() {
 
-			if ( ! is_wp_error( $response ) ) {
-				set_transient( $cache_key, $response, DAY_IN_SECONDS );
+			if ( is_array( $response ) ) {
+				$response = (object) $response;
 			}
```
Is this because WP can return both an array and an object and you want to avoid catering for both cases?

johnwatkins0 replied:
Yes, I think the `themes_api` response is likely to be an object here, but that function makes an external request and is dealing with an API version. It's not entirely transparent to me. But if we think `themes_api`, without filters, will always reliably return a stdClass object or a WP_Error, then we could remove this.

schlessera replied:
There's an old WordPress saying: "Anything that can be filtered will be filtered to be unusable" - better safe than sorry.

schlessera commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,38 @@ public function get_default_reader_themes() {
+			} else {
+				if ( empty( $response->themes ) ) {
+					$this->themes_api_error = new WP_Error(
+						'amp_themes_api_invalid_response',
```
This needs a different ID, maybe something like this:
```suggestion
						'amp_themes_api_broken_plugin_override',
```

johnwatkins0 replied:
Went with `amp_themes_api_empty_themes_array` in 89485bf0d

schlessera commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,38 @@ public function get_default_reader_themes() {
+			 * @see https://wordpress.org/support/topic/issue-during-activating-the-updated-plugins/#post-13383737
+			 */
+			if (
+				is_wp_error( $response )
+				|| ! is_object( $response )
+				|| ! property_exists( $response, 'themes' )
+				|| ! is_array( $response->themes )
```
In terms of performance and of logical progression, I think it makes more sense to start with the language construct `is_object()`. `is_object()` covers bigger ground than `is_wp_error()` (it entails that that is not true) and is faster.
```suggestion
				! is_object( $response )
				|| is_wp_error( $response )
				|| ! property_exists( $response, 'themes' )
				|| ! is_array( $response->themes )
```

schlessera replied:
Actually, I think it makes sense to cover the full happy path first, and only check for the `WP_Error` last:
```suggestion
				! is_object( $response )
				|| ! property_exists( $response, 'themes' )
				|| ! is_array( $response->themes )
				|| is_wp_error( $response )
```
All three checks from the happy path are language constructs, whereas the `is_wp_error()` is a userland function (and thus slower), and it represents the "exceptional" case.

schlessera commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,38 @@ public function get_default_reader_themes() {
+					__( 'The request for reader themes from the WordPress.org resulted in an invalid response. Please try again later or contact your host.', 'amp' )
+				);
+				return [];
+			} else {
```
The `else` is not needed here, as the `if` already ends in a `return`. This will get rid of one level of indentation.

Basically, both `if`s here are early returns, with the `set_transient()` being the main operation that is left.

schlessera replied:
Couldn't add a suggestion across the entire block, so here's what I mean as a screenshot:

![Image 2020-09-17 at 11 13 23 PM](https://user-images.githubusercontent.com/83631/93528994-8223ea80-f93b-11ea-802e-78a9e7420f2c.png)

johnwatkins0 replied:
Good point. Updated in 89a3681ec87083bd0ac28790960a5d31811e33b8

schlessera commented on tests/php/test-class-amp-reader-themes-rest-controller.php:
```diff
@@ -13,7 +13,7 @@
  * @group reader-themes
  *
- * @covers AMP_Reader_Theme_REST_Controller
+ * @coversDefaultClass \AmpProject\AmpWP\Admin\ReaderThemes
```
This seems to be wrong, the class that is tested is `AMP_Reader_Theme_REST_Controller` and the methods being tested are part of that, like `::get_items()`.

I suspect this was only meant to change the annotation type like so:
```suggestion
 * @coversDefaultClass AMP_Reader_Theme_REST_Controller
```

schlessera commented on tests/php/test-class-amp-reader-themes-rest-controller.php:
```diff
@@ -96,4 +97,43 @@ public function test_get_items() {
+	 *
+	 * @covers ::get_items
+	 */
+	public function test_get_items_with_themes_api_failure() {
```
This test actually covers two different scenarios, the header for a successful request and the header for an unsuccessful request; hence the need to do an additional `delete_transient()` even though that is already done in the `setUp()`.

Those two test scenarios should be split up into two separate tests, removing the need for an additional `delete_transient()` within the test.

johnwatkins0 replied:
Done in 437096551b6e47e5e447caa9dcba4c06bad8864f

johnwatkins0 committed:
Reorder invalid theme results check in if statement

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

johnwatkins0 committed:
Remove unnecessary else after return

johnwatkins0 committed:
Correction to coversDefaultClass

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

johnwatkins0 committed:
Break the test of themes_api success into its own method

johnwatkins0 committed:
Merge branch 'fix/5367-themes-api-result-sanitization' of https://github.com/ampproject/amp-wp into fix/5367-themes-api-result-sanitization

johnwatkins0 committed:
Use unique WP Error code for error resulting from empty themes array in themes_api

johnwatkins0 commented:
Thanks for the review, @schlessera. I have addressed all feedback.

westonruter commented on assets/src/components/themes-api-error/index.js:
```diff
@@ -0,0 +1,31 @@
+		<AMPNotice type={ NOTICE_TYPE_WARNING }>
+			<p>
+				{
+					themesAPIError
```
What about `esc_html()`?!? 😉

westonruter commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,38 @@ public function get_default_reader_themes() {
+			) {
+				$this->themes_api_error = new WP_Error(
+					'amp_themes_api_invalid_response',
+					__( 'The request for reader themes from the WordPress.org resulted in an invalid response. Please try again later or contact your host.', 'amp' )
```
```suggestion
					__( 'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.', 'amp' )
```

westonruter commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,38 @@ public function get_default_reader_themes() {
+			if ( empty( $response->themes ) ) {
+				$this->themes_api_error = new WP_Error(
+					'amp_themes_api_empty_themes_array',
+					__( 'The default reader themes cannot be displayed because a plugin is overriding the themes from WordPress.org.', 'amp' )
```
```suggestion
					__( 'The default reader themes cannot be displayed because a plugin appears to be overriding the themes response from WordPress.org.', 'amp' )
```

westonruter commented on tests/php/test-class-amp-reader-themes-rest-controller.php:
```diff
@@ -96,4 +97,48 @@ public function test_get_items() {
+
+		$response = $this->controller->get_items( new WP_REST_Request( 'GET', 'amp/v1' ) );
+		$this->assertEquals(
+			[ 'X-AMP-Theme-API-Error' => 'The default reader themes cannot be displayed because a plugin is overriding the themes from WordPress.org.' ],
```
```suggestion
			[ 'X-AMP-Theme-API-Error' => 'The default reader themes cannot be displayed because a plugin appears to be overriding the themes response from WordPress.org.' ],
```

westonruter commented on tests/php/test-class-amp-reader-themes-rest-controller.php:
```diff
@@ -96,4 +97,48 @@ public function test_get_items() {
+
+		$response = $this->controller->get_items( new WP_REST_Request( 'GET', 'amp/v1' ) );
+		$this->assertEquals(
+			[ 'X-AMP-Theme-API-Error' => 'The request for reader themes from the WordPress.org resulted in an invalid response. Please try again later or contact your host.' ],
```
```suggestion
			[ 'X-AMP-Theme-API-Error' => 'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.' ],
```

westonruter commented on tests/php/src/Admin/ReaderThemesTest.php:
```diff
@@ -104,6 +105,63 @@ public function test_get_themes() {
+		$error = $this->reader_themes->get_themes_api_error();
+		$this->assertWPError( $error );
+		$this->assertEquals(
+			'The request for reader themes from the WordPress.org resulted in an invalid response. Please try again later or contact your host.',
```
```suggestion
			'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.',
```

westonruter commented on tests/php/src/Admin/ReaderThemesTest.php:
```diff
@@ -104,6 +105,63 @@ public function test_get_themes() {
+		$error = $this->reader_themes->get_themes_api_error();
+		$this->assertWPError( $this->reader_themes->get_themes_api_error() );
+		$this->assertEquals(
+			'The default reader themes cannot be displayed because a plugin is overriding the themes from WordPress.org.',
```
```suggestion
			'The default reader themes cannot be displayed because a plugin appears to be overriding the themes response from WordPress.org.',
```

westonruter commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,38 @@ public function get_default_reader_themes() {
+				! is_object( $response )
+				|| ! property_exists( $response, 'themes' )
+				|| ! is_array( $response->themes )
+				|| is_wp_error( $response )
```
If `$response` is a `WP_Error` (and `WP_DEBUG_DISPLAY` is enabled) why not set it to be `$this->themes_api_error` here? Or at least use the message in the `amp_themes_api_invalid_response` error?

johnwatkins0 replied:
Done in 0e13feccc

westonruter commented:
This is testing well. 

When I have this plugin code active:

```php
add_filter( 'themes_api_result', function () {
	return new WP_Error( 'oops' );
} );
```

I get:

Settings Screen | Wizard
----------------|---------
![image](https://user-images.githubusercontent.com/134745/93537282-f80a6080-f8ff-11ea-9d30-6584b95d537c.png) | ![image](https://user-images.githubusercontent.com/134745/93537308-05274f80-f900-11ea-9cfd-0332c92f38b3.png)

And with this code:

```php
add_filter( 'themes_api_result', function ( $res ) {
	$res->themes = [];
	return $res;
} );
```

I get:

Settings Screen | Wizard
----------------|--------
![image](https://user-images.githubusercontent.com/134745/93537386-330c9400-f900-11ea-81fe-c7218b70bc11.png) | ![image](https://user-images.githubusercontent.com/134745/93537413-40c21980-f900-11ea-96e2-f3bcb3c05312.png)

And then as soon as I remove any such plugin code:

Settings Screen | Wizard
----------------|---------
![image](https://user-images.githubusercontent.com/134745/93537468-6818e680-f900-11ea-9edf-eca357bd182b.png) | ![image](https://user-images.githubusercontent.com/134745/93537480-7535d580-f900-11ea-8a52-2568d650576e.png)

johnwatkins0 committed:
Use error from themes_api

johnwatkins0 committed:
Revise themes_api error messages

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,45 @@ public function get_default_reader_themes() {
+				|| ! is_array( $response->themes )
+				|| is_wp_error( $response )
+			) {
+				if ( is_wp_error( $response ) ) {
+					$message = $response->get_error_message();
+				} else {
+					$message = __( 'The request for reader themes from the WordPress.org resulted in an invalid response. Please try again later or contact your host.', 'amp' );
+				}
```
```suggestion
				$message = __( 'The request for reader themes from the WordPress.org resulted in an invalid response. Please try again later or contact your host.', 'amp' );
				if ( is_wp_error( $response ) && defined( 'WP_DEBUG_DISPLAY' ) && WP_DEBUG_DISPLAY ) ) {
					$message .= ' ' . __( 'Error message:', 'amp' ) . ' ' . $response->get_error_message();
				}
```

johnwatkins0 committed:
Update error text

westonruter commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,45 @@ public function get_default_reader_themes() {
+				|| ! is_array( $response->themes )
+				|| is_wp_error( $response )
+			) {
+				if ( is_wp_error( $response ) ) {
+					$message = $response->get_error_message();
+				} else {
+					$message = __( 'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.', 'amp' );
+				}
```
```suggestion
				$message = __( 'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.', 'amp' );
				if ( is_wp_error( $response ) && defined( 'WP_DEBUG_DISPLAY' ) && WP_DEBUG_DISPLAY ) ) {
					$message .= ' ' . __( 'Error message:', 'amp' ) . ' ' . $response->get_error_message();
				}
```

johnwatkins0 replied:
Ah, I see what you mean. Thanks!

johnwatkins0 committed:
Include themes_api error message in error message if WP_DEBUG_DISPLAY is on

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter committed:
Fix syntax error and acount for no error message

westonruter commented:
Fixing tests…

westonruter committed:
Fix tests and ensure HTTP requests are mocked

schlessera commented on src/Admin/ReaderThemes.php:
```diff
@@ -210,17 +227,49 @@ public function get_default_reader_themes() {
+				|| ! is_array( $response->themes )
+				|| is_wp_error( $response )
+			) {
+				$message = __( 'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.', 'amp' );
```
```suggestion
				$message = __( 'The request for reader themes from WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.', 'amp' );
```

schlessera commented on tests/php/src/Admin/ReaderThemesTest.php:
```diff
@@ -104,6 +107,90 @@ public function test_get_themes() {
+		$error = $this->reader_themes->get_themes_api_error();
+		$this->assertWPError( $error );
+		$this->assertEquals(
+			'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.',
```
```suggestion
			'The request for reader themes from WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.',
```

schlessera commented on tests/php/src/Admin/ReaderThemesTest.php:
```diff
@@ -104,6 +107,90 @@ public function test_get_themes() {
+		$error = $this->reader_themes->get_themes_api_error();
+		$this->assertWPError( $this->reader_themes->get_themes_api_error() );
+		$this->assertStringStartsWith(
+			'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.',
```
```suggestion
			'The request for reader themes from WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.',
```

schlessera commented on tests/php/test-class-amp-reader-themes-rest-controller.php:
```diff
@@ -96,4 +101,75 @@ public function test_get_items() {
+
+		$response = $this->controller->get_items( new WP_REST_Request( 'GET', 'amp/v1' ) );
+		$this->assertEquals(
+			[ 'X-AMP-Theme-API-Error' => 'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.' ],
```
```suggestion
			[ 'X-AMP-Theme-API-Error' => 'The request for reader themes from WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.' ],
```

schlessera commented on tests/php/test-class-amp-reader-themes-rest-controller.php:
```diff
@@ -96,4 +101,75 @@ public function test_get_items() {
+		$response = $this->controller->get_items( new WP_REST_Request( 'GET', 'amp/v1' ) );
+		$headers  = $response->get_headers();
+		$this->assertArrayHasKey( 'X-AMP-Theme-API-Error', $headers );
+		$this->assertStringStartsWith( 'The request for reader themes from the WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.', $headers['X-AMP-Theme-API-Error'] );
```
```suggestion
		$this->assertStringStartsWith( 'The request for reader themes from WordPress.org resulted in an invalid response. Check your Site Health to confirm that your site can communicate with WordPress.org. Otherwise, please try again later or contact your host.', $headers['X-AMP-Theme-API-Error'] );
```

westonruter committed:
Apply suggestions from code review

Remove extraneous "the"

Co-authored-by: Alain Schlesser <alain.schlesser@gmail.com>

pierlon commented:
**QA passed**

With the plugin code:

```php
add_filter( 'themes_api_result', '__return_null' );
```
No PHP notices or errors are reported, and a warning notice is shown informing the user that their was an issue retrieving the list of reader themes:

Settings screen | Wizard screen
---|---
![image](https://user-images.githubusercontent.com/16200219/93543801-054d3c80-f94c-11ea-98fc-ef10070f0616.png) | ![image](https://user-images.githubusercontent.com/16200219/93543875-33328100-f94c-11ea-9a6e-b40670ffc563.png)

And with this code:

```php
add_filter( 'themes_api_result', function ( $res ) {
	$res->themes = [];
	return $res;
} );
```

A warning notice is shown informing the user that a plugin may be overriding the themes response:

Settings screen | Wizard screen
---|---
![image](https://user-images.githubusercontent.com/16200219/93544152-df746780-f94c-11ea-847b-9a4c8ef77c34.png) | ![image](https://user-images.githubusercontent.com/16200219/93544069-a805bb00-f94c-11ea-81b5-7cc4ab2dfffa.png)
